File: src/main/java/com/giant/common/api/doc/error/ErrorResponseDoc.java
package com.giant.common.api.doc.error;

import com.giant.common.api.type.ValidationError;
import io.swagger.v3.oas.annotations.media.Schema;
import lombok.Getter;

import java.util.List;

public class ErrorResponseDoc {
    /** 400 Bad Request - Validation Error */
    @Getter
    public static class BadRequest {
        @Schema(example = "VE")
        private final String code = null;
        @Schema(example = "입력값이 잘못되었습니다.")
        private final String message = null;

        @Schema(description = "에러내용")
        private final List<ValidationError> validationErrors = null;
    }

    /** 401 Unauthorized */
    @Getter
    public static class Unauthorized {
        @Schema(example = "UA")
        private final String code = null;
        @Schema(example = "로그인이 필요합니다.")
        private final String message = null;
    }
    /** 401 Unauthorized */
    @Getter
    public static class LoginError {
        @Schema(example = "LGE")
        private final String code = null;
        @Schema(example = "아이디 또는 비밀번호가 잘못되었습니다.")
        private final String message = null;
    }

    /** 403 Forbidden */
    @Getter
    public static class Forbidden {
        @Schema(example = "FB")
        private final String code = null;
        @Schema(example = "접근 권한이 없습니다.")
        private final String message = null;
    }

    /** 404 Not Found */
    @Getter
    public static class ResourceNotFound {
        @Schema(example = "RNF")
        private final String code = null;
        @Schema(example = "요청한 리소스를 찾을 수 없습니다.")
        private final String message = null;
    }

    /** 404 Not Found */
    @Getter
    public static class EndpointNotFound {
        @Schema(example = "ENF")
        private final String code = null;
        @Schema(example = "요청한 경로가 잘못되었습니다.")
        private final String message = null;
    }

    /** 409 Conflict */
    @Getter
    public static class DuplicateResource {
        @Schema(example = "DR")
        private final String code = null;
        @Schema(example = "이미 존재하는 항목입니다.")
        private final String message = null;
    }

    /** 409 Conflict */
    @Getter
    public static class Conflict {
        @Schema(example = "CF")
        private final String code = null;
        @Schema(example = "요청이 현재 상태와 충돌합니다.")
        private final String message = null;
    }

    /** 500 Internal Server Error */
    @Getter
    public static class InternalServerError {
        @Schema(example = "ISE")
        private final String code = null;
        @Schema(example = "서버에서 문제가 발생했습니다.")
        private final String message = null;
    }
}

File: src/main/kotlin/com/giant/common/api/doc/error/ErrorResponseDoc.kt
package com.giant.common.api.doc.error

import com.giant.common.api.type.ValidationError
import io.swagger.v3.oas.annotations.media.Schema

class ErrorResponseDoc {

    /** 400 Bad Request - Validation Error */
    class BadRequest {
        @get:Schema(example = "VE")
        val code: String? = null

        @get:Schema(example = "입력값이 잘못되었습니다.")
        val message: String? = null

        @get:Schema(description = "에러내용")
        val validationErrors: List<ValidationError>? = null
    }

    /** 401 Unauthorized */
    class Unauthorized {
        @get:Schema(example = "UA")
        val code: String? = null

        @get:Schema(example = "로그인이 필요합니다.")
        val message: String? = null
    }

    /** 401 Unauthorized */
    class LoginError {
        @get:Schema(example = "LGE")
        val code: String? = null

        @get:Schema(example = "아이디 또는 비밀번호가 잘못되었습니다.")
        val message: String? = null
    }

    /** 401 Unauthorized */
    class PasswordError {
        @get:Schema(example = "PWE")
        val code: String? = null

        @get:Schema(example = "비빌번호가 잘못되었습니다.")
        val message: String? = null
    }

    /** 403 Forbidden */
    class Forbidden {
        @get:Schema(example = "FB")
        val code: String? = null

        @get:Schema(example = "접근 권한이 없습니다.")
        val message: String? = null
    }

    /** 404 Not Found */
    class ResourceNotFound {
        @get:Schema(example = "RNF")
        val code: String? = null

        @get:Schema(example = "요청한 리소스를 찾을 수 없습니다.")
        val message: String? = null
    }

    /** 404 Not Found */
    class EndpointNotFound {
        @get:Schema(example = "ENF")
        val code: String? = null

        @get:Schema(example = "요청한 경로가 잘못되었습니다.")
        val message: String? = null
    }

    /** 409 Conflict */
    class DuplicateResource {
        @get:Schema(example = "DR")
        val code: String? = null

        @get:Schema(example = "이미 존재하는 항목입니다.")
        val message: String? = null
    }

    /** 409 Conflict */
    class Conflict {
        @get:Schema(example = "CF")
        val code: String? = null

        @get:Schema(example = "요청이 현재 상태와 충돌합니다.")
        val message: String? = null
    }

    /** 500 Internal Server Error */
    class InternalServerError {
        @get:Schema(example = "ISE")
        val code: String? = null

        @get:Schema(example = "서버에서 문제가 발생했습니다.")
        val message: String? = null
    }
}